Webapp installation and offline mode

paolobrasolin opened the pull request.

paolobrasolin committed:
Implement installation as webapp (#44) and offline mode via service worker (#157)

varkor commented on src/manifest.json:
```diff
@@ -0,0 +1,22 @@
+{
+  "short_name": "quiver",
```
Is there any reason to include a `short_name` that is the same as `name`?

paolobrasolin replied:
Not really; I'm removing it, as it's not mandatory.

I'm also removing the `scope`, which is just set to its default value, in the spirit of keeping this minimal, see https://github.com/varkor/quiver/pull/176/commits/57168be4f919b9c9cc1fdf2031c3842bd89033bf

varkor commented on src/manifest.json:
```diff
@@ -0,0 +1,22 @@
+{
+  "short_name": "quiver",
+  "name": "quiver",
+  "description": "a modern commutative diagram editor",
```
```suggestion
  "description": "A modern commutative diagram editor.",
```
I think this is the usual convention for `description`.

varkor commented on src/manifest.json:
```diff
@@ -0,0 +1,22 @@
+  "start_url": ".",
+  "scope": ".",
+  "display": "standalone",
+  "theme_color": "#000000",
```
```suggestion
  "theme_color": "#1a1a1a",
```
This is the colour used in the **quiver** logo.

varkor commented on src/index.html:
```diff
@@ -41,6 +41,13 @@
+    <link rel="manifest" href="manifest.json">
+    <!-- Service worker -->
+    <script type="module">
+        import { Workbox } from "workbox/workbox-window.prod.mjs";
```
```suggestion
        import { Workbox } from "./workbox/workbox-window.prod.mjs";
```
Without this, I get an error when I run this in localhost:
```
Uncaught TypeError: Failed to resolve module specifier "workbox/workbox-window.prod.mjs". Relative references must start with either "/", "./", or "../".
```

paolobrasolin replied:
Ah, my bad. Thanks for spotting this.

varkor commented on Makefile:
```diff
@@ -3,14 +3,28 @@
 	mv katex src/KaTeX
 
+# Vendor workbox dependencies.
+src/workbox: src/workbox/workbox-window.prod.mjs src/workbox/workbox-sw.js src/workbox/workbox-core.prod.js src/workbox/workbox-routing.prod.js src/workbox/workbox-strategies.prod.js
```
```suggestion
src/workbox: src/workbox/workbox-window.prod.mjs src/workbox/workbox-window.prod.mjs.map src/workbox/workbox-sw.js src/workbox/workbox-core.prod.js src/workbox/workbox-routing.prod.js src/workbox/workbox-strategies.prod.js
```
I also needed to add this file to avoid a 404 error.

paolobrasolin replied:
Yes, I left out the source maps because that's just a warning from the browser DevTools; however they could come in handy while debugging in production.

I'm adding the maps of all workbox dependencies (the service worker is also raising the same warning for each one), see https://github.com/varkor/quiver/pull/176/commits/d7a402fce3b07edf077a2fe3db21c21cd1f09001

varkor replied:
Ah, I see, thanks.

varkor reviewed (changes requested):
Thanks for this!

A few minor comments. I'm having trouble testing this, but it could be because I've not tested service workers before. I'm serving `src/` in localhost, opening `index.html`, then enabling the "Offline" throttling mode in the Network developer options in Chrome (or the "Offline" option in the service workers tab), then refreshing. I would expect **quiver** to continue working, but instead I get the default Chrome offline page. Is there a different way I should be testing this?

paolobrasolin committed:
Update src/manifest.json

Co-authored-by: varkor <github@varkor.com>

paolobrasolin committed:
Update src/manifest.json

Co-authored-by: varkor <github@varkor.com>

paolobrasolin committed:
Drop redundant attributes from manifest

paolobrasolin committed:
Update src/index.html

Co-authored-by: varkor <github@varkor.com>

paolobrasolin committed:
Add workbox sourcemaps download to Makefile

paolobrasolin commented:
> A few minor comments. I'm having trouble testing this, but it could be because I've not tested service workers before. I'm serving `src/` in localhost, opening `index.html`, then enabling the "Offline" throttling mode in the Network developer options in Chrome (or the "Offline" option in the service workers tab), then refreshing. I would expect **quiver** to continue working, but instead I get the default Chrome offline page. Is there a different way I should be testing this?

The first time you load the page, the SW gets installed and activated (lifecycle progress is visible in DevTools > Application > Service Workers). 

If you immediately go offline and reload the page, the service worker will restart but its lazy loaded dependencies won't be cached yet, as it wasn't caching them yet the first time it started since it had not been installed yet.

This is also why if you reload the page _before_ going offline (but after the SW started the first time) it'll work, as it had a chance to cache its own dependencies on the second run.

Not really super intuitive. However this is an edge due just to my hurried implementation which basically works only after the first visit to the app. This is solvable by precaching the assets at install time, instead of capturing the request and captuiring them at runtime. To avoid any other possible edge case, it'd be worth it to generate a manifest of all assets (including images, JS, etc) and cache it. I'll think of a clever way of generating it (without overcomplicating the build) and get back to you.

In the meantime, if you reload the page before going offline you should be able to see the SW working properly, @varkor.

paolobrasolin committed:
Drop handmade service worker

paolobrasolin committed:
Use workbox-build to build service worker

paolobrasolin committed:
Add note about build order in makefile

paolobrasolin commented:
Ok, @varkor. I have the solution. I switched to using `workbox-build`; it's a CLI tool to build simpl-ish service worker automatically, and it includes manifest generation and dependency bundling.

I think this really is the best way, but it is a **`node`** CLI tool. This could be annoying, so I'm keeping everything (tiny code file, big lock file) inside a new `service-worker/` folder to avoid polluting the repo. I'm also calling `node` via `docker` from the `Makefile` to make it dead simple to use (maybe you prefer `nix`, or `nvm`, but that's easily fixed).

(Maybe someday you will need a `node` build for the whole app, and then it'll make sense to keep everything at the top level.)

This time it should actually work. The asset pre-fetching is performed during the installation of the service worker, so no weird edge cases are possible. You can see the requests happening in the Network tab of the Chrome DevTools if you're curious (MathJax is by far the heaviest dependency).

Lmk if that works for you, and what you think. Thanks!

varkor commented:
(Just to let you know that I'm currently on holiday, and will get back to this as soon as I can.)

varkor commented on .gitignore:
```diff
@@ -1 +1,4 @@
 src/KaTeX/*
+src/icon-*.png
+src/workbox-*
+src/service-worker.js*
```
```suggestion
src/service-worker.js*

```

varkor commented on service-worker/.gitignore:
```diff
@@ -0,0 +1 @@
+node_modules/
```
```suggestion
node_modules/

```

varkor reviewed (changes requested):
@paolobrasolin: I've taken another look (apologies for the wait). I think I'm happy with this approach. Would it be possible to switch to using nvm rather than Docker? It seems like a more lightweight solution (particularly on macOS). After that, I think it should be good to go :)

varkor commented:
@paolobrasolin: have you had a chance to look at this again?

paolobrasolin committed:
Switch from docker to nvm for node

paolobrasolin committed:
Add trailing newline to gitignores

paolobrasolin commented:
Hi @varkor! Sorry for the wait, I got sidetracked.

I just rebased, switched to `nvm` in the `Makefile` (it should still be shell/platform agnostic), and added the trailing newlines in the `.gitignore`s. 👍 

Sidenote: I'm using `node` 20 in the `Makefile` as it's the active LTS release, but anything `>=16` will do (and is required by the lock file).

varkor commented:
Thanks for updating! I will try to take a look at this later today, or by the end of the week if I don't get to it then.

varkor committed:
Tweak comment

varkor commented:
Thanks, this looks great! I'll probably push a new minor version of **quiver** soon, given how useful this feature is. Thank you for implementing this feature!

varkor commented:
@paolobrasolin: I've noticed a small issue (which I should have tested for earlier). It appears that, after the service worker is installed, it will not be updated. For instance, if a user visits the page for the first time, so the service worker is installed, then the page is updated, the user will see the version of the page that they saw when the initially visited. In Chrome, hard reloading fixes the issue temporarily, but later (non-hard) refreshes of the page give the old version. Furthermore, if the user goes offline afterwards, they will see the old version of the page, rather than the latest version (even with a hard reload).

Do you know if there's a straightforward fix for this? (I'll investigate more later, but I imagine you might know the answer immediately.)

paolobrasolin commented:
Uhm. Thanks for letting me know @varkor! I expected this to be handled automatically, tbh.
Let me unpack how this is supposed to work, so maybe we can figure out the issue.

`service-worker/build.js` is a blueprint to generate a standardized service worker with the `workbox` library.
`make src/service-worker.js` simply executes it, producing `src/service-worker.js` as an output.
`src/service-worker.js` is the actual code that gets registered as a service worker and then executed by the browser.

Registration is handled in `index.html` via the convenience library `workbox-window`. Not much to see there.

The content of `service-worker/build.js` amounts to
```js
const { generateSW } = require("workbox-build");
generateSW({globDirectory: "../src/", globPatterns: ["**/*"], swDest: "../src/service-worker.js"})
```
which simply generates a manifest of all files matching `../src/**/*` and feeds it to the `precacheAndRoute` function.

What is the manifest? This array, where I'm omitting the revision number of the files:
```js
[
  { url: "arrow.js", revision: "..." },
  { url: "bezier.js", revision: "..." },
  { url: "dom.js", revision: "..." },
  { url: "ds.js", revision: "..." },
  { url: "icon.png", revision: "..." },
  { url: "icons/about.svg", revision: "..." },
  { url: "icons/centre-view.svg", revision: "..." },
  { url: "icons/delete.svg", revision: "..." },
  { url: "icons/deselect-all.svg", revision: "..." },
  { url: "icons/hide-grid.svg", revision: "..." },
  { url: "icons/pullback-checked.svg", revision: "..." },
  { url: "icons/pullback-unchecked.svg", revision: "..." },
  { url: "icons/redo.svg", revision: "..." },
  { url: "icons/reset-zoom.svg", revision: "..." },
  { url: "icons/save.svg", revision: "..." },
  { url: "icons/select-all.svg", revision: "..." },
  { url: "icons/shortcuts.svg", revision: "..." },
  { url: "icons/show-hints.svg", revision: "..." },
  { url: "icons/show-queue.svg", revision: "..." },
  { url: "icons/undo.svg", revision: "..." },
  { url: "icons/var-pullback-checked.svg", revision: "..." },
  { url: "icons/var-pullback-unchecked.svg", revision: "..." },
  { url: "icons/zoom-in.svg", revision: "..." },
  { url: "icons/zoom-out.svg", revision: "..." },
  { url: "index.html", revision: "..." },
  { url: "main.css", revision: "..." },
  { url: "manifest.json", revision: "..." },
  { url: "parser.js", revision: "..." },
  { url: "quiver-blue.svg", revision: "..." },
  { url: "quiver.js", revision: "..." },
  { url: "quiver.svg", revision: "..." },
  { url: "service-worker.js.map", revision: "..." },
  { url: "tests/parser.tex", revision: "..." },
  { url: "ui.js", revision: "..." },
  { url: "workbox-2b403519.js.map", revision: "..." },
];
```

In our case the revision numbers are just file hashes computed when `service-worker/build.js` is executed.
**So this is the first important point: you need to re`make src/service-worker.js` for every release, or the service worker won't be able to invalidate the cache automatically.**

What is the [`precacheAndRoute` function](https://developer.chrome.com/docs/workbox/modules/workbox-precaching#method-precacheAndRoute)? Well, it just performs a `precache` (immediately fetches and caches assets upon initialization of the service worker) and an `addRoute` (sets up a router which will answer to future fetch requests with a cache-first strategy) for every file of the given manifest.

This means that unless you do a force-reload, the cache will always prevail unless it's invalidated. (This explains the behaviour you're observing when force-reloading.)

**Now, a _very_ important aspect of the lifecycle of a service worker is that its default behaviour is to replace older versions of itself only when they are not being executed anymore.** As in: you need to close ALL quiver tabs coming from a given host. You can see this in the Chrome inspector under `Application > Service workers`:

![image](https://github.com/varkor/quiver/assets/2499655/d230f818-d8fa-483b-8ada-6a3555a9d103)

Above you see I am serving `quiver` locally, have a single open tab (client), and version `#1746` is waiting for `#1744` to die in order to replace it. You can click `skipWaiting` to force that to happen, or close and reopen the tab. After that you get:

![image](https://github.com/varkor/quiver/assets/2499655/19b1eb30-407e-4b73-a478-ae326f56fc95)

Now the new version is running, meaning that the manifest was updated, old cache was invalidated and the new assets are being used.

---

While writing all of the above I was also checking the behaviour of `quiver`, and it seems to behave as expected.

So I can think of two problems that may be causing what you see:
* you might not be re`make`ing `src/service-worker.js`, so the manifest does not get updated
* you might not be triggering the restart of the service worker, so the old service worker does not get updated (keeping a tab open is enough to prevent it)

Both are compatible with the force-reload behaviour you describe and could be causing the issue.

Final note: it is possible to configure the service worker to just `skipWaiting`. It's usually not recommended to do it lightly since it can break stuff if one does background computation or persists state with their service workers, but we're just doing some precaching here, so it should be fine. On the other hand, while it's confusing for a developer, the end user probably wouldn't notice the difference: the worst case with the current behaviour is that they run the updated version only after closing and reopening the tab (or the browser).

Sorry for the wall of text, I hope this clarifies things a bit. 😄

varkor commented:
@paolobrasolin: thanks, that explanation helps a lot. I had overlooked both of the problems you identified, and things seem to be working as expected now that my misunderstanding has been corrected.

> On the other hand, while it's confusing for a developer, the end user probably wouldn't notice the difference: the worst case with the current behaviour is that they run the updated version only after closing and reopening the tab (or the browser).

Yes, I agree, it's probably not an issue in practice for users.

---

I have one more novice question: I'd like to place the generated files in a subdirectory (e.g. `src/Workbox/`) so they don't clutter `src/`. This isn't an issue for the `workbox-window` files. However, when I tried placing `service-worker.js` and the other generated `workbox-<ID>.js` file in a subdirectory, there was a precaching error (not surprisingly). I tried fixing this by adding:
```
modifyURLPrefix: {
    "": "../",
  }
```
to the `generateSW`, which addressed the error (the paths in `service-worker.js` now being correct), but the page would no longer work offline. Is it possible to place `service-worker.js` in a different directory from the content it is supposed to cache, like this? It's not a big problem if it's not convenient to do; it would just make the folder structure a little cleaner. Thanks in advance :)

varkor commented:
Note to myself: the issue is that the service worker must be in the root directory, due to service worker scoping (https://github.com/GoogleChrome/workbox/issues/2530). However, this means I need to work out the best way to push these generated files to GitHub Pages, because the file name is not configurable, so I cannot just hardcode the filenames (https://github.com/GoogleChrome/workbox/issues/2637). This is not a problem, as I can just use an environment variable.

The bigger issue at the moment is that I need to ensure sensible service worker behaviour across https://q.uiver.app/ and https://q.uiver.app/dev.

varkor commented:
The update is now live :)
